src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp

```diff
@@ -0,0 +1,534 @@
+constexpr unsigned int w64 = 64;
+constexpr unsigned int max_cmd_args = 8;
+constexpr unsigned int fw_9 = 9;
+
+class launcher
+{
+ public:
+  // Public accessor for the singleton instance
+  static launcher& getInstance()
+  {
+    static launcher instance;
+    return instance;
+  }
+
+  // Public members
+  std::array<char, str_sz_s> m_name = {0};
+  bool m_debug = false;
+
+  // Delete copy constructor and assignment operator
+  launcher(const launcher&) = delete;
+  launcher& operator=(const launcher&) = delete;
+  launcher(launcher&& other) noexcept = delete;
+  launcher& operator=(launcher&& other) noexcept = delete;
+
+ private:
+  launcher() = default;
+  ~launcher() = default;
+};
+
+/*
+ * This function template appends a given value of any type to the specified
+ * output string stream using the stream insertion operator.
+ */
+template <typename T>
+void log_format(std::ostringstream& oss, const T& t)
+{
+  /* Warning: implicitly decay an array into a pointer
+  * The function is explicitly designed to handle any type, including arrays
+  * Behavior of array-to-pointer decay as safe in this context
+  */
+  oss << t; /* NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay,
+  hicpp-no-array-decay) */
+}
+
+/*
+ * This function template appends a given value of any type to the specified
+ * output string stream and recursively processes additional values if provided.
+ * It terminates when there are no more values to process.
+ */
+template <typename T, typename... Args>
+void log_format(std::ostringstream& oss, const T& t, const Args&... args)
+{
+  /* Warning: implicitly decay an array into a pointer
+  * The function is explicitly designed to handle any type, including arrays
+  * Behavior of array-to-pointer decay as safe in this context
+  */
+  oss << t; /* NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay,
+  hicpp-no-array-decay) */
+  log_format(oss, args...);
+}
+
+/*
+ * Function to trace the error log
+ */
+template <typename... Args>
+void log_e(const Args&... args)
+{
+  std::ostringstream oss;
+  oss << "[" << launcher::getInstance().m_name.data() << "] E: ";
+  log_format(oss, args...);
+  std::cout << oss.str() << std::endl;
+}
+
+/*
+ * Function to trace the fatal log
+ */
+template <typename... Args>
+void log_f(const Args&... args)
+{
+  std::ostringstream oss;
+  oss << "[" << launcher::getInstance().m_name.data() << "] F: ";
+  log_format(oss, args...);
+  throw std::runtime_error(oss.str() + ". Aborted!\n");
+}
+
+/*
+ * Function to trace the debug log
+ */
+template <typename... Args>
+void log_d(const Args&... args)
+{
+  if (launcher::getInstance().m_debug) {
+    std::ostringstream oss;
+    oss << "[" << launcher::getInstance().m_name.data() << "] D: ";
+    log_format(oss, args...);
+    std::cout << oss.str() << std::endl;
+  }
+}
+
+/*
+ * thread safe wrapper of putenv()
+ */
+void putenv_t(std::string& new_entry)
+{
+  std::lock_guard<std::mutex> lock(env_mutex);
```

Comment:
Looks like env_mutex is used exclusively within this function. So do this and remove the global.
```suggestion
  static std::mutex mutex;
  std::lock_guard lock(mutex);
```

Reply:
Done